prov/verbs/src/verbs_ep.c

```diff
@@ -183,7 +183,7 @@ ssize_t vrb_post_send(struct vrb_ep *ep, struct ibv_send_wr *wr, uint64_t flags)
 	}
 
 	if (!ep->sq_credits || !ep->peer_rq_credits) {
-		cq = container_of(ep->util_ep.rx_cq, struct vrb_cq, util_cq);
+		cq = container_of(ep->util_ep.tx_cq, struct vrb_cq, util_cq);
```

Comment:
I think the problem is slightly worse.  If sq_credits == 0, then we need to read transmit completions to free send credits.  If peer_rq_credits == 0, then we need to read receive completions to look for new credits sent by the peer.

Maybe the entire if statement needs to be split into 2?

Reply:
>If peer_rq_credits == 0, then we need to read receive completions to look for new credits sent by the peer.
that's RXM though.  how would you do that without replicating RXM code from here?

Reply:
Hmm... good point.  I think the way to fix this, so that the code reflects what's happening is:
```
if (!ep->peer_rq_credits)
    goto freectx;

if (!ep->sq_credits) {
    cq = ... tx_cq ...   // your fix
    vrb_flush_cq(cq)

    if (!ep->sq_credits)
        goto freectx;
}
```

Reply:
yeah.  leave peer_rq_credits to normal progress call.

Reply:
Side question, why does vrb_wr_consumes_recv macro include IBV_WR_RDMA_WRITE_WITH_IMM?  write with immediate should not consume a RX buffer, immediate data shows up in CQ not RX buffer.  What am I missing?

Reply:
Verbs write with immediate consumes RX buffer. It's the OFI write w/imm that doesn't.